src/Disks/ObjectStorages/MetadataStorageMetrics.h

```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include <Disks/DiskType.h>
+#include <Common/CurrentMetrics.h>
+#include <Common/ProfileEvents.h>
+
+namespace DB
+{
+
+struct MetadataStorageMetrics
+{
+    const ProfileEvents::Event directory_created = ProfileEvents::end();
+    const ProfileEvents::Event directory_removed = ProfileEvents::end();
+
+    CurrentMetrics::Metric directory_map_size = CurrentMetrics::end();
+
+    template <typename ObjectStorage, MetadataStorageType metadata_type>
+    static MetadataStorageMetrics create()
+    {
+        return MetadataStorageMetrics{};
```

Comment:
do we want to have a default value that returns `end` for all metrics or would we like to have a compile error?

maybe instead of instantiations we could have `if constexpr` here and on last else have `static_assert(false, ....)`?

Reply:
Nice suggestion.
I've tried to implement it, but there is a difficulty: metadata storage compatibility with the object storage is checked at runtime, not compile time.
There is a storage for which `plain_rewritable` metadata type does not make sense – `WebObjectStorage`, yet a user may create the following spec:
```
type = object_storage,
object_type = web,
metadata_type = plain_rewritable
```
An exception will be thrown: https://github.com/ClickHouse/ClickHouse/blob/master/src/Disks/ObjectStorages/ObjectStorageFactory.cpp#L80-L86

Yet, it may still be compiled with incompatible (or not yet supported HDFS) storages: https://pastila.nl/?002628e9/a50d61928f45a114c0dd402b79b18818#Mo8U3U9kpWaZCEEsjgWTaA==

Reply:
okay thanks for trying it out!